# **Gradient Descent**


*  Gradient Descent is the technique to train the machine learining model, its a iterative way of choose the best paramters values for 'm', 'b' which minimize the error (Cost function).
*italicized text*





In case of linear regression the cost fuction we have is MSE- mean aquare error, given by:-


$$J(w,b) = \frac{1}{n} \sum_{i=1}^{n} (y_i - (wx_i + b))^2$$
where,$$(wx_i + b) = predicted, ({pred}_i)$$ <br>
<br>
      w - weight(slope) <br>
      b - bias(intercept) <br>



*   Signifies the rate of change in y
 (the target) for every one-unit increase in x
 (the feature).
*   Represents the value of the target variable when the input 𝑥𝑖
 is exactly 0.



*   $y = wx + b$, this is the linear regression equation.

The goal is to find the values of w and b such a way that we get the "best fit line".





# Working principal of Gradient Descent:


*  Suppose we have dataset:
 x (Input) = [1,2,3,4]  and y (Target) = [2,3,5,6]

*   We want to fit these values, our equation of line is 𝑦=𝑤𝑥+𝑏

In Gradient Descent we choose initial values(randomly).

*  Initial Guess: $w = 0$, $b = 0$ (starting point).
*  Learning Rate ($\alpha$): suppose we take $0.05$
*  Number of points ($n$): 4.


Steps in Gradient Descent are as followed:















# Step 1:

We need to find the Gradients: <br>
We need to take partial derivatves of MSE , J(w, b) w.r.t w and b in order to find these Gradients.<br>


Gradients of $w$: $D_w = \frac{2}{n} \sum -x_i(y_i -
\text{pred}_i)$  and <br>
Gradients of $b$: $D_b = \frac{2}{n} \sum -(y_i - \text{pred}_i)$ <br>



# Step 2:
Now calculate the prediction and error with w = 0 and b = 0 as initial values.

when these values are putted in 𝑦=𝑤𝑥+𝑏, 𝑦 becomes 0 meaning the prediction are 0 for any values for x. ad hence the error will be , Error (𝑦𝑖−pred𝑖) = [2,3,5,6]





# Step 3:

Now, find the values Gradient,

$D_b = \frac{2}{n} \sum -(y_i - \text{pred}_i)$<br>

(𝑦𝑖−pred𝑖) = [2,3,5,6] = 16 (sum of errors)

$D_b = \frac{-2}{4} $ (16) <br>

$D_b = -8$ <br>

Similarly find for w
$D_w = \frac{2}{n} \sum -x_i(y_i -
\text{pred}_i)$ <br>
$D_w = -23.5$


# Step 4: Update the parameters

$w = w_0 - (\alpha \cdot D_w)$ <br>
$b = b_0 - (\alpha \cdot D_b)$<br>
$\alpha \ is\ 0.05$ <br>
$w_0$ = 0 <br>
$b_0$ = 0 <br>

  new value of w = 1.175<br>
  and b is 0.4 <br>

so, the eqaution of line will be now y = 1.175x + 0.4<br>

this process will repeart untill minima is reached, minimum error eventually 0.





# Python Implementation:


In [8]:
import numpy as np

def line(x, w, b): # (best fit line)
    return w * x + b

def MSE(x, y, w, b): # Cost Function
    n = len(x)
    predictions = line(x, w, b)
    mse = (1/n) * np.sum((y - predictions)**2)
    return mse

def gradient_descent(x, y, w, b, learning_rate): # Gradient Descent
    n = len(x)
    predictions = line(x, w, b)
    error = y - predictions
    dw = (2/n) * np.sum(-x * error)  # Calculate gradients
    db = (2/n) * np.sum(-error)
    new_w = w - (learning_rate * dw) # Update parameters
    new_b = b - (learning_rate * db)
    return new_w, new_b

In [13]:
x_data = np.array([1, 2, 3, 4])
y_data = np.array([2, 3, 5, 6])

# Initials
w, b = 0.0, 0.0
lr = 0.05

for i in range(1000):
    w, b = gradient_descent(x_data, y_data, w, b, lr)

    if i % 100 == 0:
        print(f"Step {i}: mse = {MSE(x_data, y_data, w, b):.4f}")

print(f"Final Optimized Weights: w={w:.2f}, b={b:.2f}")

Step 0: mse = 0.5522
Step 100: mse = 0.0500
Step 200: mse = 0.0500
Step 300: mse = 0.0500
Step 400: mse = 0.0500
Step 500: mse = 0.0500
Step 600: mse = 0.0500
Step 700: mse = 0.0500
Step 800: mse = 0.0500
Step 900: mse = 0.0500
Final Optimized Weights: w=1.40, b=0.50


Initial MSE: $\frac{74}{4} = \mathbf{18.5}$

1. Calculating the Gradients ($D_w$ and $D_b$) For the Bias($b$) <br>
$$D_b = \frac{2}{n} \sum -(y - \text{pred}) = \frac{2}{4} \times (-16) = \mathbf{-8.0}$$<br>

2. For the Weight ($w$):<br>
multiply each error by its $x$ value first:<br>
Point 1: $-(1 \times 2) = -2$<br>
Point 2: $-(2 \times 3) = -6$<br>
Point 3: $-(3 \times 5) = -15$<br>
Point 4: $-(4 \times 6) = -24$<br>
$Sum: $-47$$
$D_w = \frac{2}{n} \times (\text{Sum}) = \frac{2}{4} \times (-47) = \mathbf{-23.5}$$


3. The Update parameters<br>
New $w$: $0 - (0.05 \times -23.5) = \mathbf{+1.175}$<br>
New $b$: $0 - (0.05 \times -8.0) = \mathbf{+0.40}$

4. New MSE with $y = 1.175x + 0.4$:<br>
For $x=1, y_{pred} = 1.575 \rightarrow Error^2 = (2 - 1.575)^2 = 0.1806$ <br> For $x=2, y_{pred} = 2.75 \rightarrow Error^2 = (3 - 2.75)^2 = 0.0625$ <br> For $x=3, y_{pred} = 3.925 \rightarrow Error^2 = (5 - 3.925)^2 = 1.1556$ <br> For $x=4, y_{pred} = 5.1 \rightarrow Error^2 = (6 - 5.1)^2 = 0.8100$<br>

MSE: $\frac{0.1806 + 0.0625 + 1.1556 + 0.8100}{4} = \mathbf{0.5522}$


1. Purpose: To find the optimal parameters (Weights/Bias) that result in the
lowest possible error.

2. Use Case: Essential for models where you can't solve for the answer in one single math step (like Deep Learning or large-scale Linear Regression).

3. The Goal: Minimize the Cost Function (MSE), moving from a random guess to a "Best Fit."

4. The Component $D_w$: Measures how sensitive the error is to the Slope of the line.

5. The Component $D_b$: Measures how sensitive the error is to the Height of the line.

6. The Learning Rate ($\alpha$)The "Step Size": Determines how much of the gradient we actually use to update the weight. <br>
**Too Low:** High precision but extremely slow. It might take 10,000 steps to reach the bottom. <br>
**Too High**: The model "overshoots" the valley, jumps to the other side, and the error starts increasing (Divergence).

7. The Logic (The Update Rule)The Formula: $w = w - (\alpha \times \text{Gradient})$.The "Flipper": Because we subtract, a negative slope pushes the weight up, and a positive slope pushes the weight down. <br>  The Magnitude: Large error = Large Gradient = Big steps. Small error = Small Gradient = Small steps.